### PR TITLE
Feature/remove get url

### PR DESCRIPTION
**Pull request by p0psicles**
The idea is to rip all the PoliceSession stuff out of this. And to do a refactor and new features in separate PR's. This PR should be ready to merge, when proven stable in not showing any regression using the Session objects.

- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

For discussion purposes i've already created the poc. But this is far from production ready.

### Refactoring make_session / get_url
- [x] Remove all the get_url calls for providers with the PolicedSession()
- [x] Remove all the get_url calls for rest of medusa
- [x] Remove the _request_defaults method from the MedusaSession
- [x] Replace make_session's
- [x] ~~Implement the TextSession, JsonSession and ContentSession where needed. These should only be some edge cases, but we need to got through all calls anyway.~~
- [x] Implemented session helpers, to be used with the MedusaSession or MedusaSafeSession. The following helpers are available: get_content, get_json and get_text.

#### Exception handling
- [x] Implement exception handling
A subclass MedusaSafeSession() is available with the most basic exception handling.
A response object is returned if available. Else None is returned. Making sure the MedusaSafeSession can be safely dropped in.

#### TODO:
- [x] Fix tokyotosho

### CODE REVIEW

**Comment by fernandog** [comment]
@p0psicles merged develop and fixed conflicts. please check if everything is ok

**Comment by fernandog** [comment]
tested with manual search and snatched some eps. all good

**Comment by p0psicles** [comment]
Tnx. But i'm afraid this PR will need allot more testing. And it will be very hard to catch all exceptional situations because we don't have any unittests for those.

Like for example, allot of changes won't cause any issues in a sunny-day scenario. Like a json provider, that is giving back a json. But we should test what happens if the provider is having issues, and is giving back an error (404) page. Also should this PR be reviewed how we handle these kind of situations, as we sometime returned None, when à 'exception' occurred. So some decisions need to be made there.

I've at least started the ground work, so we can now easily make those decisions and just as easy implement them.

**Comment by fernandog** [comment]
@p0psicles like this right?

```
2017-05-03 17:07:15 DEBUG    SEARCHQUEUE-MANUAL-257655 :: [T411] :: [b4a8a54] Traceback (most recent call last):
  File "C:\Users\fernando\Documents\Medusa\medusa\search\core.py", line 547, in search_providers
    down_cur_quality, manual_search, manual_search_type)
  File "C:\Users\fernando\Documents\Medusa\medusa\providers\generic_provider.py", line 215, in find_search_results
    items_list += self.search(search_string, ep_obj=episode)
  File "C:\Users\fernando\Documents\Medusa\medusa\providers\torrent\json\t411.py", line 85, in search
    if not self.login():
  File "C:\Users\fernando\Documents\Medusa\medusa\providers\torrent\json\t411.py", line 219, in login
    response = self.session.post(self.urls['login_page'], data=login_params).json()
AttributeError: 'NoneType' object has no attribute 'json'
```

File: medusa/__main__.py [hunk 1]
```diff
@@ -983,6 +983,10 @@ def initialize(self, console_logging=True):
                 load_provider_setting(app.CFG, provider, 'bool', 'enable_daily', 1)
                 load_provider_setting(app.CFG, provider, 'bool', 'enable_backlog', provider.supports_backlog)
                 load_provider_setting(app.CFG, provider, 'bool', 'enable_manualsearch', 1)
+                load_provider_setting(app.CFG, provider, 'bool', 'enable_api_hit_cooldown', 0)
```
Comment (p0psicles) [inline medusa/__main__.py]: Note remove these.

File: medusa/__main__.py [hunk 2]
```diff
@@ -1508,6 +1512,8 @@ def save_config():
                     'api_key', 'username',
                     'search_mode', 'search_fallback',
                     'enable_daily', 'enable_backlog', 'enable_manualsearch',
+                    'enable_api_hit_cooldown', 'enable_daily_request_reserve',
+                    'api_hit_limit', 'enable_daily_request_reserve'
```
Comment (p0psicles) [inline medusa/__main__.py]: Note remove.

File: medusa/helper/throttle.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            start = time()
+            while not self.consume():
+                if not logged:
+                    if first_pass:
+                        log.debug('Throttling {func}'.format(func=name))
+                first_pass = False
+            finish = time()
+            log.debug('Throttled for {x} seconds'.format(x=finish - start))
+            return func(*args, **kwargs)
+        return consumer
```
Comment (p0psicles) [inline medusa/helper/throttle.py]: Remove entire module

File: medusa/providers/generic_provider.py [hunk 4]
```diff
@@ -94,7 +94,7 @@ def __init__(self, name):
         self.public = False
         self.search_fallback = False
         self.search_mode = None
-        self.session = make_session()
+        self.session = MedusaSession(hooks=[cloudflare])
```
Comment (p0psicles) [inline medusa/providers/generic_provider.py]: @labrys what if we subclass to MedusaSafeSession and add exception handling to it? Then we can safely transition to the session objects. And replace them with MedusaSession later, one by one.

File: medusa/session/custom.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+class ThrottledSession(MedusaSession):
+    """A Throttled Session that rate limits requests."""
+
+    def __init__(self, throttle, **kwargs):
+        """Initialize the class."""
+        super(ThrottledSession, self).__init__(**kwargs)
+        self.throttle = throttle
+        if self.throttle:
+            self.request = self.throttle(super(ThrottledSession, self).request)
```
Comment (p0psicles) [inline medusa/session/custom.py]: Remove module

File: medusa/session/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+class PolicedRequestScoreExceeded(PolicedRequestException):
+    """Police Request Score Exception."""
+
+
+class PolicedRequestDailyExceeded(PolicedRequestException):
+    """Police Request Exception for exceeding the reserved daily search limit."""
+
+
+class PolicedRequestInvalidConfiguration(PolicedRequestException):
+    """Invalid or incomplete configuration provided."""
```
Comment (p0psicles) [inline medusa/session/exceptions.py]: Remove module

File: views/config_providers.mako [hunk 7]
```diff
@@ -213,6 +213,39 @@ $('#config-components').tabs();
...
+                        <div class="field-pair">
+                                <label>
+                                    <span class="component-title">Daily search reserved api hits</span>
+                                    <span class="component-desc">
+                                        <input type="number" min="0" step="1" name="${cur_newznab_provider.get_id()}_daily_reserve_calls" value="${cur_newznab_provider.daily_reserve_calls}" class="form-control input-sm input75"/>
+                                        <p>Reserve this amount of api hits for the daily/rss search. Use 0 to disable.</p>
+                                    </span>
+                                </label>
+                        </div><!-- usenet retention -->
+                        % endif
```
Comment (p0psicles) [inline views/config_providers.mako]: Revert these changes

**Comment by p0psicles** [comment]
```2017-07-03 14:08:50 ERROR SEARCHQUEUE-DAILY-SEARCH :: [HorribleSubs] :: [5af1bf7] Failed parsing provider. Traceback: 'Traceback (most recent call last):\n File "/opt/medusa/medusa/providers/torrent/html/horriblesubs.py", line 111, in parse\n download_url = row.find(\'td\', class_=\'dl-type hs-magnet-link\').a.get(\'href\')\nAttributeError: \'NoneType\' object has no attribute \'a\'\n' Traceback (most recent call last): File "/opt/medusa/medusa/providers/torrent/html/horriblesubs.py", line 111, in parse download_url = row.find('td', class_='dl-type hs-magnet-link').a.get('href') AttributeError: 'NoneType' object has no attribute 'a'
2017-07-03 14:08:50 ERROR SEARCHQUEUE-DAILY-SEARCH :: [HorribleSubs] :: [5af1bf7] Failed parsing provider. Traceback: 'Traceback (most recent call last):\n File "/opt/medusa/medusa/providers/torrent/html/horriblesubs.py", line 111, in parse\n download_url = row.find(\'td\', class_=\'dl-type hs-magnet-link\').a.get(\'href\')\nAttributeError: \'NoneType\' object has no attribute \'a\'\n' Traceback (most recent call last): File "/opt/medusa/medusa/providers/torrent/html/horriblesubs.py", line 111, in parse download_url = row.find('td', class_='dl-type hs-magnet-link').a.get('href') AttributeError: 'NoneType' object has no attribute 'a' ```

File: medusa/clients/nzb/nzbget.py [hunk 8]
```diff
@@ -145,7 +145,8 @@ def sendNZB(nzb, proper=False):
                     if not nzb.provider.login():
                         return False
 
-                    data = nzb.provider.get_url(nzb.url, returns='content')
+                    # TODO: Check if this needs exception handling
+                    data = nzb.provider.session(nzb.url).content
                     if data is None:
                         return False
```
Comment (labrys) [inline medusa/clients/nzb/nzbget.py]: You forgot the `.get`.  Also its better written as follows:
```

File: medusa/clients/nzb/nzbget.py [hunk 9]
```diff
@@ -145,7 +145,8 @@ def sendNZB(nzb, proper=False):
                     if not nzb.provider.login():
                         return False
 
-                    data = nzb.provider.get_url(nzb.url, returns='content')
+                    # TODO: Check if this needs exception handling
+                    data = nzb.provider.session(nzb.url).content
```
Comment (labrys) [inline medusa/clients/nzb/nzbget.py]: This is wrong for several reasons.  First the `.get` for the call is missing.  Secondly we need to create a session specifically for `nzbget`.  Additionally, if the response is needed in the future for further error handling, it is no longer provided.  A better approach would be:
```python
response = self.session.get(nzb.url)
data = response.content
```

File: medusa/helpers/__init__.py [hunk 10]
```diff
@@ -1574,11 +1533,13 @@ def get_disk_space_usage(disk_path=None, pretty=True):
 
 def get_tvdb_from_id(indexer_id, indexer):
 
-    session = make_session()
+    content_session = ContentSession()
+    session = MedusaSession()
```
Comment (labrys) [inline medusa/helpers/__init__.py]: You have two sessions here, you should only have one.  All the content_session, text_session, etc need to be removed.  They reduce functionality with no gain.

File: medusa/clients/nzb/nzbget.py (hunk 9 above)
Comment (p0psicles) [inline medusa/clients/nzb/nzbget.py]: I agree

File: medusa/helpers/__init__.py (hunk 10 above)
Comment (p0psicles) [inline medusa/helpers/__init__.py]: But I'd like to have some sugar coating, to be able to easily 'drop-in', where we expect content, or json. Should I go with the get_json() then?

**Comment by p0psicles** [comment]
Okay wrapping up this PR, as i'd like to prevent a 6th rebase. So please give your feedback, or we'll merge shortly.

**Comment by labrys** [comment]
On the road.. will take another look at it tonight if we don't get back to late.

**Pull request merged**